pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/IdenticalCatchBranchesRule.java

```diff
@@ -0,0 +1,157 @@
+            if (result.isEmpty()) {
+                result.add(newEquivClass(stmt));
+                continue;
+            }
+
+            boolean isNewClass = true;
+            for (List<ASTCatchStatement> equivClass : result) {
+                if (areEquivalent(stmt, equivClass.get(0))) {
+                    equivClass.add(stmt);
+                    isNewClass = false;
+                    break;
+                }
+            }
+
+            if (isNewClass) {
+                result.add(newEquivClass(stmt));
+            }
+        }
+
+        return result;
+    }
+
+
+    private List<ASTCatchStatement> newEquivClass(ASTCatchStatement stmt) {
+        // Each equivalence class is sorted by document order
+        List<ASTCatchStatement> result = new ArrayList<>(2);
+        result.add(stmt);
+        return result;
+    }
+
+
+    // Gets the representation of the set of catch statements as a single multicatch
+    private String getCaughtExceptionsAsString(ASTCatchStatement stmt) {
+
+        StringBuilder sb = new StringBuilder();
+
+        final String delim = " | ";
+        for (ASTType type : stmt.getCaughtExceptionTypeNodes()) {
+            sb.append(type.getTypeImage()).append(delim);
+        }
+
+        // remove the last delimiter
+        sb.replace(sb.length() - 3, sb.length(), "");
+        return sb.toString();
+    }
+
+
+    @Override
+    public Object visit(ASTTryStatement node, Object data) {
+
+        List<ASTCatchStatement> catchStatements = node.getCatchStatements();
+        Set<List<ASTCatchStatement>> equivClasses = equivalenceClasses(catchStatements);
+
+        for (List<ASTCatchStatement> identicalStmts : equivClasses) {
+            if (identicalStmts.size() > 1) {
+                String identicalBranchName = getCaughtExceptionsAsString(identicalStmts.get(0));
+
+                // By convention, lower catch blocks are collapsed into the highest one
+                // The first node of the equivalence class is thus the block that should be transformed
+                for (int i = 1; i < identicalStmts.size(); i++) {
+                    addViolation(data, identicalStmts.get(i), new String[]{identicalBranchName, });
+                }
+            }
+        }
+
+        return super.visit(node, data);
+    }
+
+    /**
+     * Checks whether two nodes has same subtree.
+     * Note, it wouldn't be sensitive to changes in the exception variable name.
+     *
+     * @param node1
+     *             the first node to check
+     * @param node2
+     *             the second node to check
+     * @param exceptionName1
+     *             the first exception variable name
+     * @param exceptionName2
+     *             the second exception variable name
+     * @return <code>ture</code> if two nodes has same subtree, otherwise <code>false</code>
+     */
+    private boolean hasSameSubTree(Node node1, Node node2, String exceptionName1, String exceptionName2) {
```

Comment:
When we tackle the AST refactoring for PMD 7.0.0, we can think about adding such a method directly to ASTCatchStatement - allowing to compare a CatchStatement to another one (Not sure, if this would be the behavior of equals, or whether we introduce a compare/same method instead).

Then the above `areEquivalent` method would look like this:

```java
private boolean areEquivalent(ASTCatchStatement st1, ASTCatchStatement st2) {
    return st1.isSame(st2);
}
```